pkg/squashfs/squashfs.go

```diff
@@ -300,36 +289,310 @@ func squashFuse(squashFile, extractDir string) (*exec.Cmd, error) {
+	k.mutex.Lock()
+	defer k.mutex.Unlock()
+
+	if mounted, err := isMountedAtDir(squashFile, extractDir); mounted && err == nil {
+		log.Debugf("[%s] %s already mounted -> %s", k.Name(), squashFile, extractDir)
+		return true, nil
+	} else if err != nil {
+		return false, err
+	}
+
+	cmd, err := squashFuse(squashFile, extractDir)
+	if err != nil {
+		return false, err
+	}
+
+	log.Debugf("squashFuse mounted (%d) %s -> %s", cmd.Process.Pid, squashFile, extractDir)
+	if err := cmd.Process.Release(); err != nil {
+		return false, errors.Errorf("Failed to release process %s: %v", cmd, err)
+	}
+	return true, nil
+}
+
+// ExtractSingleSquashPolicy - extract squashfile to extractDir
+func ExtractSingleSquashPolicy(squashFile, extractDir string, policy *ExtractPolicy) error {
+	const initName = "init"
+	if policy == nil {
+		return errors.Errorf("policy cannot be nil")
+	}
+
+	// avoid taking a lock if already initialized (possibly premature optimization)
+	if !policy.initialized {
+		policy.mutex.Lock()
+		// We may have been waiting on the initializer. If so, then the policy will now be initialized.
+		// if not, then we are the initializer.
+		if !policy.initialized {
+			defer policy.mutex.Unlock()
+			defer func() {
+				policy.initialized = true
+			}()
+		} else {
+			policy.mutex.Unlock()
+		}
+	}
+
 	err := os.MkdirAll(extractDir, 0755)
 	if err != nil {
 		return err
 	}
 
-	if mounted, err := maybeKernelSquashMount(squashFile, extractDir); err == nil && mounted {
-		return nil
-	} else if err != kernelSquashMountFailed {
+	fdest, err := filepath.Abs(extractDir)
+	if err != nil {
 		return err
 	}
 
-	cmd, err := squashFuse(squashFile, extractDir)
-	if err == nil {
-		if err := cmd.Process.Release(); err != nil {
-			return errors.Errorf("Failed to release process %s: %v", cmd, err)
+	if policy.initialized {
+		if err, ok := policy.Excuses[initName]; ok {
+			return err
+		}
+		mounted, err := policy.Extractor.Mount(squashFile, fdest)
+		if err == nil {
+			if !mounted {
+				return errors.Errorf("%s.Mount(%s, %s) returned unmounted with err=nil",
+					policy.Extractor.Name(), squashFile, fdest)
+			}
+			return nil
 		}
-		return nil
-	} else if err != squashNotFound {
 		return err
 	}
-	if p := which("unsquashfs"); p != "" {
-		log.Debugf("Extracting %s -> %s with unsquashfs -f -d %s %s", extractDir, squashFile, extractDir, squashFile)
-		cmd := exec.Command("unsquashfs", "-f", "-d", extractDir, squashFile)
-		cmd.Stdout = os.Stdout
-		cmd.Stderr = os.Stderr
-		cmd.Stdin = nil
-		return cmd.Run()
+
+	// At this point we are the initialzer
+	if policy.Excuses == nil {
+		policy.Excuses = map[string]error{}
+	}
+
+	if policy.Extractors == nil || len(policy.Extractors) == 0 {
+		policy.Excuses[initName] = errors.Errorf("policy had no extractors")
+		return policy.Excuses[initName]
+	}
+
+	var extractor SquashExtractor
+	var mounted bool
+	allExcuses := []string{}
+	for _, extractor = range policy.Extractors {
```

Comment:
this block runs only when we have the initializer lock, it tries all Extractors and sets policy.Extractor on first success.